DM-8536 Add postage stamp cutouts and fastavro

mtpatter opened the pull request:
Add postage stamps to alert sending and receiving.  Use the fastavro module for much faster alert serialization when using the Avro format.

mtpatter committed:
upgrade librdkafka for Docker image

mtpatter committed:
use fastavro module for serializing

The fastavro module serializes ~4x faster and deserializes ~10x faster than the standard avro module.

mtpatter committed:
update schemas for avro

Add new avro schemas for including postage stamp cutout and for use with fastavro module.

mtpatter committed:
Add postage stamps to alerts in sender

mtpatter committed:
Enable postage stamp collection in consumer

mtpatter committed:
Fix consumer for receiving data when decode=False

If decode=False, does not encode to Avro format. Data structures that don't conform to an Avro schema (e.g., dict, list, set, tuple) can be received.

mtpatter committed:
Add Readme instructions for postage stamps

SimonKrughoff commented on bin/printAlertStream.py:
```diff
@@ -9,9 +9,34 @@
+def msg_text(message):
+    """Remove postage stamp cutouts from an alert message.
+    """
+    message_text = {k: message[k] for k in message if k not in ['cutoutDifference', 'cutoutTemplate']}
```
In several places there are hard coded keys like this.  It makes me a little concerned that if the schema changes, it will break.  Is there a way to either figure out these keys from the schema or to have it configurable in a parent class so that we only have to change things in one place?

mtpatter replied:
I will have to think about this. (Also, the schema files here are only temporarily hardcoded in the command line scripts- later we can use a schema registry to distribute schemas.) If using Avro and the writer schema needs changing, we can update it without breaking the reader, even if the reader does not have the updated schema. The readers can have their own schemas that can be significantly different from the writers (different key names, etc.) as long as the data fits (which will be a nice feature for filtering down the line) so I don't think we want to go down the line of enforcing too much learning from the original schema.

SimonKrughoff commented on README.md:
```diff
@@ -71,3 +69,28 @@ $ docker-compose down
+
+**On Avro**
+
+Using the Avro serialization format to enforce a schema for the reader/writer is turned on by default but can be turned off by setting encode=False and decode=False in send/poll.
```
I'm curious why this requires a rebuild.  It seems like you could add an --encode/--decode argument to each of the producers and consumers.  Is it just then you could potentially get things out of sync?

mtpatter replied:
You're right-it doesn't, I will just add that.  Much easier.  👍

SimonKrughoff reviewed (approved):
This looks really excellent.  I had a couple of comments.  

One sort of off topic question: `monitorAlertStream.py` defaults to `decode=False` [here](https://github.com/lsst-dm/alert_stream/blob/tickets/DM-8536/bin/monitorAlertStream.py#L35).  On the other hand, `printeAlertStream.py` defaults to `decode=True` [here](https://github.com/lsst-dm/alert_stream/blob/tickets/DM-8536/bin/printAlertStream.py#L70).  It would be good to be consistent.

mtpatter commented:
> One sort of off topic question: monitorAlertStream.py defaults to decode=False here. On the other hand, printAlertStream.py defaults to decode=True here. It would be good to be consistent.

The sender and printer both default to encoding/decoding which I think I want to keep turned on by default to enforce the schema formatting, so that is consistent.  The monitor doesn't care about message content because it just logs when it has finished listening to all available alerts, and I want it to be as fast as possible so it doesn't interpret any messages or need to decode anything.

mtpatter committed:
Make stamps / no stamps flags mutually exclusive.

Without this, the last flag wins.

mtpatter committed:
Make Avro encoding/decoding command line arguments

mtpatter committed:
Rename consumer scripts

These scripts can be used for listening to any Kafka topic stream, not just alerts.

mtpatter committed:
Update Readme
